File: Task2/Code/object_oriented.cpp
//------------------------------------------------------------------------------
// object_oriented.cpp - содержит функции обработки Объектно-ориентированных языков
//------------------------------------------------------------------------------

#include "object_oriented.h"

// от 1 до 3 для наследования
Random ObjectOriented::rnd_inheritance(1, 3);

//------------------------------------------------------------------------------
// Ввод параметров Объектно-ориентированного языка из потока
void ObjectOriented::In(ifstream &ifst) {
    int key_of_inheritance;
    string name;
    ifst >> name >> key_of_inheritance >> popularity >> year;
    name_len = name.size();
    switch (key_of_inheritance) {
        case (single):
            inheritance = single;
            break;
        case (multiple):
            inheritance = multiple;
            break;
        case (interfaces):
            inheritance = interfaces;
            break;
    }
}

// Случайный ввод параметров Объектно-ориентированного языка
void ObjectOriented::InRnd() {
    name_len = rnd_len.GetInt();
    int key = rnd_inheritance.GetInt();
    switch (key) {
        case (single):
            inheritance = single;
            break;
        case (multiple):
            inheritance = multiple;
            break;
        case (interfaces):
            inheritance = interfaces;
            break;
    }
    popularity = rnd_popularity.GetDouble();
    year = rnd_year.GetInt();
}

//------------------------------------------------------------------------------
// Вывод параметров Объектно-ориентированного языка в поток
void ObjectOriented::Out(ofstream &ofst) {
    ofst << "It is Object oriented programming language: number of char. = "
         << name_len << ",  inheritance = ";
    switch (inheritance) {
        case (single):
            ofst << "single";
            break;
        case (multiple):
            ofst << "multiple";
            break;
        case (interfaces):
            ofst << "interfaces";
            break;
    }
    ofst << ".   \t\t\tPopularity = " << popularity
         << "%, year = " << year
         << ", quotient = " << Quotient() << "\n";
}

//------------------------------------------------------------------------------
// Вычисление частного Объектно-ориентированного языка
double ObjectOriented::Quotient() {
    return year / double(name_len);
}

File: Task3/Code/objectOriented.py
import random
from language import Language

# Класс Объектно-ориентированного языка программирования
class ObjectOriented(Language):
    def __init__(self):
        self.name_len = 0  # длина названия языка
        self.inheritance = 0  # наследование
        self.popularity = 0  # популярность
        self.year = 0  # год создания

    # Ввод данных в объект класса из строки файла
    def input(self, str):
        self.name_len = int(len(str[1]))
        self.inheritance = int(str[2])
        self.popularity = float(str[3])
        self.year = int(str[4])

    # Ввод данных в объект класса случайным образом
    def inputrnd(self):
        self.name_len = random.randint(1, 20)
        self.inheritance = random.randint(1, 3)
        self.popularity = round(random.random() * (100 - 1) + 1, 3)
        self.year = random.randint(1957, 2020)
        pass

    # Вывод данных объекта класса в файл
    def output(self, ostr):
        if self.inheritance == 1:
            self.inheritance = "single.\t"
        elif self.inheritance == 2:
            self.inheritance = "multiple."
        else:
            self.inheritance = "interfaces."
        ostr.write(
            f"It is Object oriented programming language: number of char. = {self.name_len},  "
            f"inheritance = {self.inheritance}    \t\t\tPopularity = {self.popularity}%, "
            f"year = {self.year}, "
            f"quotient = {self.quotient()}\n")
        pass

    # Подсчет частного для данного объекта класса языка программирования
    def quotient(self):
        return round(self.year / float(self.name_len), 3)
